### PR TITLE
bpo-43897: Implement AST validation for Pattern Matching

### PR DESCRIPTION
**Pull request by isidentical**
https://bugs.python.org/issue43897

### CODE REVIEW

**Review by brandtbucher — COMMENTED** [review]
Hi @isidentical!

I think I missed this in the beta-freeze scramble. Any plans to update it for the new AST?

**Comment by isidentical** [comment]
> I think I missed this in the beta-freeze scramble. Any plans to update it for the new AST?

I might take a look at it in the upcoming weeks though I don't think I would have enough capacity to catch up with beta 2 (9 days from now). If you have time, I could give access to my fork if not I'll try to make it ready for beta 3.

**Comment by brandtbucher** [comment]
Beta 3 is fine!

**Comment by isidentical** [comment]
Hey @brandtbucher! I've updated the PR with the new AST form, sorry it took a bit long. I didn't have much free time after the AST layout change, and needed to catch up with the new nodes first. Just ported all the missing pieces, there are still some bugs that I can see (for example some of the old parts of the validator code just quit when error happens, but doesn't decrement the recursion limit. Since I didn't touch those and the patch was already big I just skipped them now, might fix later) but nothing too big. Please let me know any missing stuff that I should handle.

File: Lib/test/test_ast.py [hunk 1]
```diff
@@ -1507,6 +1507,152 @@ def test_stdlib_validates(self):
...
+    def test_match_validation_pattern(self):
+        name_x = ast.Name('x', ast.Load())
+        for pattern in self._MATCH_PATTERNS:
+            with self.subTest(ast.dump(pattern, indent=4)):
+                node = ast.Match(
+                    subject=name_x,
+                    cases = [
+                        ast.match_case(
+                            pattern=pattern,
+                            body = [ast.Expr(ast.Constant(...))]
```
Comment (brandtbucher) [inline Lib/test/test_ast.py]: Took me a sec to see what was happening here. Maybe clearer as `ast.Pass()`?
```suggestion
                            body = [ast.Pass()]
```

File: Lib/test/test_ast.py [hunk 2]
```diff
@@ -1507,6 +1507,152 @@ def test_stdlib_validates(self):
...
+        ast.MatchMapping(
+            [constant_true, ast.Starred(ast.Name('lol', ast.Load()), ast.Load())],
+            [pattern_x, pattern_1],
+            rest='legit'
+        ),
+        ast.MatchMapping(
+            [constant_true, ast.Starred(ast.Name('lol', ast.Load()), ast.Load())],
+            [pattern_x, pattern_1],
+            rest='legit'
+        ),
```
Comment (brandtbucher) [inline Lib/test/test_ast.py]: This looks the exact same as the pattern above, right?

```suggestion
```

File: Python/ast.c [hunk 3]
```diff
@@ -447,7 +448,23 @@ validate_pattern_match_value(struct validator *state, expr_ty exp)
...
+            PyObject *literal = exp->v.Constant.value;
+            if (PyLong_CheckExact(literal) || PyFloat_CheckExact(literal) ||
+                PyBytes_CheckExact(literal) || PyComplex_CheckExact(literal) ||
+                PyUnicode_CheckExact(literal)) {
+                return 1;
+            }
+            PyErr_SetString(PyExc_ValueError,
+                            "unexpected constant inside of a literal pattern");
+            return 0;
+        }
```
Comment (brandtbucher) [inline Python/ast.c]: I don't think these braces are needed, right?

```suggestion
        case Constant_kind:
            /* Ellipsis and immutable sequences are not allowed.
               For True, False and None, MatchSingleton() should
               be used */
            if (!validate_expr(state, exp, Load)) {
                return 0;
            }
            PyObject *literal = exp->v.Constant.value;
            if (PyLong_CheckExact(literal) || PyFloat_CheckExact(literal) ||
                PyBytes_CheckExact(literal) || PyComplex_CheckExact(literal) ||
                PyUnicode_CheckExact(literal)) {
                return 1;
            }
            PyErr_SetString(PyExc_ValueError,
                            "unexpected constant inside of a literal pattern");
            return 0;
```

File: Python/ast.c [hunk 4]
```diff
@@ -489,51 +509,97 @@ validate_pattern(struct validator *state, pattern_ty p)
...
-            ret = 1;
+            ret = validate_patterns(state, p->v.MatchSequence.patterns, /*star_ok=*/1);
             break;
-        case MatchMapping_kind:
-            // TODO: check "rest" target name is valid
+        case MatchMapping_kind: {
             if (asdl_seq_LEN(p->v.MatchMapping.keys) != asdl_seq_LEN(p->v.MatchMapping.patterns)) {
                 PyErr_SetString(PyExc_ValueError,
                                 "MatchMapping doesn't have the same number of keys as patterns");
                 return 0;
```
Comment (brandtbucher) [inline Python/ast.c]:
```suggestion
                ret = 0;
                break;
```

File: Python/ast.c [hunk 5]
```diff
@@ -489,51 +509,97 @@ validate_pattern(struct validator *state, pattern_ty p)
...
             break;
         case MatchSequence_kind:
-            // TODO: Validate all subpatterns
-            // return validate_patterns(state, p->v.MatchSequence.patterns);
-            ret = 1;
+            ret = validate_patterns(state, p->v.MatchSequence.patterns, /*star_ok=*/1);
             break;
-        case MatchMapping_kind:
-            // TODO: check "rest" target name is valid
+        case MatchMapping_kind: {
```
Comment (brandtbucher) [inline Python/ast.c]: These braces aren't needed either, right?

```suggestion
        case MatchMapping_kind:
```

File: Python/ast.c [hunk 6]
```diff
@@ -489,51 +509,97 @@ validate_pattern(struct validator *state, pattern_ty p)
...
+                    break;
+                }
             }
-            // TODO: Validate all subpatterns
-            // ret = validate_patterns(state, p->v.MatchMapping.patterns);
-            ret = 1;
+
+            ret = validate_patterns(state, p->v.MatchMapping.patterns, /*star_ok=*/0);
             break;
+        }
```
Comment (brandtbucher) [inline Python/ast.c]: See above.

```suggestion
```

File: Python/ast.c [hunk 7]
```diff
@@ -489,51 +509,97 @@ validate_pattern(struct validator *state, pattern_ty p)
...
-            ret = 1;
+
+            expr_ty cls = p->v.MatchClass.cls;
+            while (1) {
+                if (cls->kind == Name_kind) {
+                    break;
+                } else if (cls->kind == Attribute_kind) {
+                    cls = cls->v.Attribute.value;
+                    continue;
+                } else {
```
Comment (brandtbucher) [inline Python/ast.c]:
```suggestion
                }
                else if (cls->kind == Attribute_kind) {
                    cls = cls->v.Attribute.value;
                    continue;
                }
                else {
```

File: Python/ast.c [hunk 8]
```diff
@@ -489,51 +509,97 @@ validate_pattern(struct validator *state, pattern_ty p)
...
-            ret = 1;
+
+            ret = validate_patterns(state, p->v.MatchMapping.patterns, /*star_ok=*/0);
             break;
+        }
         case MatchClass_kind:
             if (asdl_seq_LEN(p->v.MatchClass.kwd_attrs) != asdl_seq_LEN(p->v.MatchClass.kwd_patterns)) {
                 PyErr_SetString(PyExc_ValueError,
                                 "MatchClass doesn't have the same number of keyword attributes as patterns");
                 return 0;
```
Comment (brandtbucher) [inline Python/ast.c]:
```suggestion
                ret = 0;
                break;
```

File: Python/ast.c [hunk 9]
```diff
@@ -489,51 +509,97 @@ validate_pattern(struct validator *state, pattern_ty p)
...
             }
-            // TODO: Restrict cls lookup to being a name or attribute
             if (!validate_expr(state, p->v.MatchClass.cls, Load)) {
                 return 0;
```
Comment (brandtbucher) [inline Python/ast.c]:
```suggestion
                ret = 0;
                break;
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]
Thanks, this looks really good! I just have a few comments:

**Comment by brandtbucher** [comment]
> Just ported all the missing pieces, there are still some bugs that I can see (for example some of the old parts of the validator code just quit when error happens, but doesn't decrement the recursion limit. Since I didn't touch those and the patch was already big I just skipped them now, might fix later) but nothing too big.

Sorry, didn't see this. Feel free to ignore those suggestions, then!

I can find time to go back over those and fix the `"_"` name validation issue sometime if you're too busy. Just let me know.

**Comment by isidentical** [comment]
> I can find time to go back over those and fix the "_" name validation issue sometime if you're too busy. Just let me know.

What do you mean by the `_` validation (example case would be welcome)? By the way feel free to apply style suggestions by yourself since it is quite hard to deal with that sort of diffs in GitHub interface :(

**Comment by brandtbucher** [comment]
> What do you mean by the `_` validation (example case would be welcome)?

Patterns like `MatchMapping(..., rest="_")`, `MatchStar("_")`, and `MatchAs(..., name="_")` should be rejected, but aren't.

Also, `MatchStar` is allowed as a top-level pattern, but shouldn't be. I'll just merge this and fix those, the remaining recursion depth bugs, and remove some outdated comments in a new PR today (it will be easier to backport to 3.10 that way).

**Comment by brandtbucher** [comment]
Thanks again! Sorry I kept dropping the ball on this, haha.

**Pull request merged**